yeonjae/programmers-2/땅따먹기.js

```diff
@@ -0,0 +1,23 @@
+function solution(land) {
+  const tempLand = [...land];
+  let answer = 0;
+
+  for (let i = 1; i < tempLand.length; i++) {
+    for (let j = 0; j < 4; j++) {
+      tempLand[i][j] += Math.max(
+        ...tempLand[i - 1].filter((v, idx) => j !== idx)
+      );
+      answer = Math.max(tempLand[i][j], answer);
+    }
+  }
+  answer = tempLand[tempLand.length - 1];
+
+  return Math.max(...answer);
+}
```

Comment:
```suggestion
function solution(land) {
  const tempLand = [...land];

  for (let i = 1; i < tempLand.length; i++) {
    for (let j = 0; j < 4; j++) {
      tempLand[i][j] += Math.max(
        ...tempLand[i - 1].filter((v, idx) => j !== idx)
      );
    }
  }
  
  return Math.max(...tempLand.pop());
}
```
- 풀이잘봤습니다
- for문 내에서 Math.max를 재할당하더라도 결국 모든 for문이 끝나면 tempLand의 마지막 요소를 재할당 하고 있어서 for문에서 계속 값을 넣더라도 무의미한 로직입니다 제거되도 정상 실행됩니다.
- 또한 answer는 처음에 숫자 0으로 초기화했는데, for문 종료 후에는 배열을 재할당해주고 있습니다. 처음 초기화해준 값과 다른 타입을 할당해주고 있어서 적절하지 않은 것 같습니다.
- 마지막으로 지금 혼란을 주는 answer는 차라리 없어도 문제 없을 것 같습니다 마지막에 templand의 마지막 요소를 가져오는 pop을 활용해도 좋을 것 같습니다.
- 아래는 reduce를 활용한 풀이입니다.
```js
function solution(land) {
  const getMaxScore = (prevScores, curScore, idx) => {
    const prevMaxScore = Math.max(...prevScores.filter((_, i) => i !== idx));

    return curScore + prevMaxScore;
  };

  const cumulativeScores = land.reduce(
    (acc, cur) => {
      return cur.map((curScore, idx) => getMaxScore(acc, curScore, idx));
    },
    [0, 0, 0, 0]
  );

  return Math.max(...cumulativeScores);
}
```

Reply:
코멘트 확인했습니다. reduce를 활용한 풀이도 참고하겠습니다 훨씬 깔끔해지네요!